conda/core/index.py

```diff
@@ -371,64 +465,61 @@ def fetch_repodata(url, cache_dir=None, use_cache=False, session=None):
             log.debug("Using cached repodata for %s at %s. Timeout in %d sec",
                       url, cache_path, timeout)
-            return read_local_repodata(cache_path)
+            return read_local_repodata(cache_path, url, schannel, priority)
 
         log.debug("Locally invalidating cached repodata for %s at %s", url, cache_path)
 
     try:
         assert url is not None, url
-        fetched_repodata = fetch_repodata_remote_request(session, url,
-                                                         mod_etag_headers.get('_etag'),
-                                                         mod_etag_headers.get('_mod'))
+        repodata = fetch_repodata_remote_request(session, url,
+                                                 mod_etag_headers.get('_etag'),
+                                                 mod_etag_headers.get('_mod'))
     except Response304ContentUnchanged:
         log.debug("304 NOT MODIFIED for '%s'. Updating mtime and loading from disk", url)
         touch(cache_path)
-        return read_local_repodata(cache_path)
+        return read_local_repodata(cache_path, url, schannel, priority)
 
     with open(cache_path, 'w') as fo:
-        json.dump(fetched_repodata, fo, indent=2, sort_keys=True, cls=EntityEncoder)
+        json.dump(repodata, fo, indent=2, sort_keys=True, cls=EntityEncoder)
+    process_repodata(repodata, url, schannel, priority)
+    write_pickled_repodata(cache_path, repodata)
+    return repodata
 
-    return fetched_repodata or None
 
-
-def _collect_repodatas_serial(use_cache, urls):
+def _collect_repodatas_serial(use_cache, tasks):
     # type: (bool, List[str]) -> List[Sequence[str, Option[Dict[Dist, IndexRecord]]]]
     session = CondaSession()
-    repodatas = [(url, fetch_repodata(url, use_cache=use_cache, session=session))
-                 for url in urls]
+    repodatas = [(url, fetch_repodata(url, schan, pri,
+                                      use_cache=use_cache,
+                                      session=session))
+                 for url, schan, pri in tasks]
     return repodatas
 
 
-def _collect_repodatas_concurrent(executor, use_cache, urls):
-    futures = tuple(executor.submit(fetch_repodata, url, use_cache=use_cache,
-                                    session=CondaSession()) for url in urls)
-    repodatas = [(u, f.result()) for u, f in zip(urls, futures)]
+def _collect_repodatas_concurrent(executor, use_cache, tasks):
+    futures = tuple(executor.submit(fetch_repodata, url, schan, pri,
+                                    use_cache=use_cache,
+                                    session=CondaSession())
+                    for url, schan, pri in tasks)
+    repodatas = [(t[0], f.result()) for t, f in zip(tasks, futures)]
     return repodatas
 
 
-def _collect_repodatas(use_cache, urls):
-    # TODO: there HAS to be a way to clean up this logic
+def _collect_repodatas(use_cache, tasks):
+    repodatas = executor = None
     if context.concurrent:
```

Comment:
FYI, I have `context.concurrent` defaulting to `False`.  I was seeing some issues filed related to it.  Nothing we can't fix, but the safest strategy was making it configurable and turning it off.  It's definitely something Ray and MikeS might want to experiment with though (and help us work out any bugs in the process).